File: 566.reshape-the-matrix.java
/*
 * @lc app=leetcode id=566 lang=java
 *
 * [566] Reshape the Matrix
 */

// @lc code=start
class Solution {
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        int originalRowSize = mat.length;
        int originalColSize = mat[0].length;

        if (originalColSize * originalRowSize != r * c) {
            return mat;
        }

        int[][] res = new int[r][c];
        int rowTracker = 0;
        int colTracker = 0;

        for (int rowIdx = 0; rowIdx < originalRowSize; rowIdx++) {
            for (int colIdx = 0; colIdx < originalColSize; colIdx++) {
                if (colTracker == c) {
                    rowTracker++;
                    colTracker = 0;
                }

                res[rowTracker][colTracker] = mat[rowIdx][colIdx];
                colTracker++;
                
            }
        }

        return res;
    }
}
// @lc code=end

File: 566.reshape-the-matrix.py
#
# @lc app=leetcode id=566 lang=python3
#
# [566] Reshape the Matrix
#

# @lc code=start
class Solution:
    def matrixReshape(self, mat: List[List[int]], r: int, c: int) -> List[List[int]]:
        
        original_rows = len(mat)
        original_cols = len(mat[0])
        
        # Early exit if reshaping is not possible
        if original_rows * original_cols != r * c:
            return mat
        
        reshaped_mat = []
        flat_list = []
        for row in mat:
            flat_list.extend(row)
        
        for i in range(0, len(flat_list), c):
            reshaped_mat.append(flat_list[i:i+c])
        
        return reshaped_mat


# @lc code=end

